### PR TITLE
generic connector with code review changes

### PR DESCRIPTION
**Pull request by krishnilak**

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/pom.xml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+ ~ specific language governing permissions and limitations
+ ~ under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>identity-outbound-oauth-oauth2</artifactId>
+        <groupId>org.wso2.carbon.identity.outbound.auth.oauth2</groupId>
+        <version>1.0-SNAPSHOT</version>
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/pom.xml]: 1.0.0-SNAPSHOT

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        </identity.framework.package.import.version.range>
+        <oauth2.generic.version.range>[0.0,1)</oauth2.generic.version.range>
+        <oltu.version.range>[1.0,2)</oltu.version.range>
+        <apache.commons.version.range>[2.6,3)</apache.commons.version.range>
+        <javax.servlet.version.range>[2.6,3)</javax.servlet.version.range>
+        <apache.commons.logging.version.range>[1.2,2)</apache.commons.logging.version.range>
+        <osgi.framework.version.range>[1.8,2)</osgi.framework.version.range>
+        <json.version.range>[3.0,4)</json.version.range>
+    </properties>
+</project>
```
Comment (GDRDABARERA) [inline pom.xml]: new line

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/pom.xml [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                <inherited>true</inherited>
+                <configuration>
+                    <encoding>UTF-8</encoding>
+                    <source>1.8</source>
+                    <target>1.8</target>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/pom.xml]: new line

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/* Oauth2GenericAuthenticator supports federating authentication with External Oauth IDP s from WSO2 IAM */
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: put correct class level comments

**Review by GDRDABARERA — CHANGES_REQUESTED** [review]

File: .idea/.name [hunk 5]
```diff
@@ -0,0 +1 @@
+identity-outbound-oauth-oauth2
```
Comment (GDRDABARERA) [inline .idea/.name]: Can we remove these files

File: .idea/.gitignore [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
+/shelf/
+/workspace.xml
```
Comment (GDRDABARERA) [inline .idea/.gitignore]: Can we remove these files

File: .idea/compiler.xml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+        <outputRelativeToContentRoot value="true" />
+        <module name="org.wso2.carbon.identity.application.authenticator.oauth2" />
+      </profile>
+    </annotationProcessing>
+    <bytecodeTargetLevel>
+      <module name="identity-outbound-oauth-oauth2" target="1.7" />
+      <module name="org.wso2.carbon.identity.application.authenticator.oauth2" target="1.8" />
+    </bytecodeTargetLevel>
+  </component>
+</project>
```
Comment (GDRDABARERA) [inline .idea/compiler.xml]: remove .idea files

File: .idea/encodings.xml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="Encoding">
+    <file url="file://$PROJECT_DIR$/components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java" charset="UTF-8" />
+  </component>
+</project>
```
Comment (GDRDABARERA) [inline .idea/encodings.xml]: Remove these .idea files

**Review by GDRDABARERA — CHANGES_REQUESTED** [review]

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 9]
```diff
@@ -0,0 +1,547 @@
...
+        }
+
+        String stateToken = generateState();
+        this.stateToken = stateToken;
+
+        try {
+            Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+            String clientId = authenticatorProperties.get(Oauth2GenericAuthenticatorConstants.CLIENT_ID);
+            String callbackUrl = authenticatorProperties.get(Oauth2GenericAuthenticatorConstants.CALLBACK_URL);
+
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Remove line space

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 10]
```diff
@@ -0,0 +1,547 @@
...
+            context.setContextIdentifier(stateToken);
+
+            String state = stateToken + "," + Oauth2GenericAuthenticatorConstants.OAUTH2_LOGIN_TYPE;
+            OAuthClientRequest authzRequest = OAuthClientRequest.authorizationLocation(authorizationEP)
+                    .setClientId(clientId).setResponseType("code").setRedirectURI(callbackUrl)
+                    .setState(state)
+                    .buildQueryMessage();
+
+            if (logger.isDebugEnabled()) {
+                logger.debug("authzRequest");
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Join this two comments to one

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 11]
```diff
@@ -0,0 +1,547 @@
...
+        try {
+            Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+            String clientId = authenticatorProperties.get(Oauth2GenericAuthenticatorConstants.CLIENT_ID);
+            String clientSecret = authenticatorProperties.get(Oauth2GenericAuthenticatorConstants.CLIENT_SECRET);
+            String redirectUri = authenticatorProperties.get(Oauth2GenericAuthenticatorConstants.CALLBACK_URL);
+            Boolean basicAuthEnabled =
+                    Boolean.parseBoolean(
+                            authenticatorProperties.get(Oauth2GenericAuthenticatorConstants.IS_BASIC_AUTH_ENABLED));
+            String code = getAuthorizationCode(request);
+
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Remove space
space

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 12]
```diff
@@ -0,0 +1,547 @@
...
+            if (logger.isDebugEnabled()) {
+                logger.debug("buildClaims");
+            }
+
+            Map<ClaimMapping, String> claims = new HashMap<>();
+
+            for (Map.Entry<String, Object> entry : userInfoJson.entrySet()) {
+                claims.put(
+
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Remove line space and format

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 13]
```diff
@@ -0,0 +1,547 @@
...
+                        ClaimMapping.build(entry.getKey(), entry.getKey(), null, false), entry.getValue().toString());
+
+                if (logger.isDebugEnabled()
+                        && IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.USER_CLAIMS)) {
+                    logger.debug("Adding claim mapping : " + entry.getKey() + " <> " + entry.getKey() + " : "
+                            + entry.getValue());
+                }
+
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Remove line space

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 14]
```diff
@@ -0,0 +1,547 @@
...
+            }
+            if (StringUtils
+                    .isBlank(context.getExternalIdP().getIdentityProvider().getClaimConfig().getUserClaimURI())) {
+                context.getExternalIdP().getIdentityProvider().getClaimConfig()
+                        .setUserClaimURI(Oauth2GenericAuthenticatorConstants.EMAIL);
+            }
+            String subjectFromClaims = FrameworkUtils
+                    .getFederatedSubjectFromClaims(context.getExternalIdP().getIdentityProvider(), claims);
+            if (subjectFromClaims != null && !subjectFromClaims.isEmpty()) {
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Shall we use StringUtils.isBlank it checks for both null and empty

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 15]
```diff
@@ -0,0 +1,547 @@
...
+        OAuthClientRequest tokenRequest = null;
+        String token;
+        String tokenResponseStr;
+        try {
+            String state = this.stateToken;
+            tokenRequest =
+                    buidTokenRequest(tokenEndPoint, clientId, clientSecret, state, code, redirectUri, basicAuthEnabled);
+            tokenResponseStr = sendRequest(tokenRequest.getLocationUri());
+
+            logger.info(tokenResponseStr);
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Is it okay to log this as info, does this has sensitive infor, make this audit

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 16]
```diff
@@ -0,0 +1,547 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/***
+ * Oauth2GenericAuthenticator supports federating authentication with External Oauth IDP s from WSO2 IAM
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Put a fullstop after the comment

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 17]
```diff
@@ -0,0 +1,547 @@
...
+ */
+public class Oauth2GenericAuthenticator extends AbstractApplicationAuthenticator
+        implements FederatedApplicationAuthenticator {
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java:68]: Add enough audit logs to all the classes

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 18]
```diff
@@ -0,0 +1,547 @@
...
+            }
+            context.getSubject().setUserAttributes(claims);
+
+        } else {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Decoded json object is null");
+            }
+            throw new AuthenticationFailedException("Decoded json object is null");
+        }
+
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Remove line

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 19]
```diff
@@ -0,0 +1,547 @@
...
+                                                  String state, String code, String redirectUri,
+                                                  Boolean basicAuthEnabled) throws ApplicationAuthenticatorException {
+
+        OAuthClientRequest tokenRequest;
+        try {
+            if (!basicAuthEnabled) {
+                tokenRequest = OAuthClientRequest.tokenLocation(tokenEndPoint).setClientId(clientId)
+                        .setClientSecret(clientSecret).setGrantType(GrantType.AUTHORIZATION_CODE).setCode(code)
+                        .setRedirectURI(redirectUri)
+                        .setParameter("state", state).buildQueryMessage();
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Put "state" to a constant and "code" too

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 20]
```diff
@@ -0,0 +1,547 @@
...
+            } else {
+                tokenRequest = OAuthClientRequest.tokenLocation(tokenEndPoint).setClientId(clientId)
+                        .setClientSecret(clientSecret).setGrantType(GrantType.AUTHORIZATION_CODE).setCode(code)
+                        .setParameter("state", state).buildQueryMessage();
+                String base64EncodedCredential = new String(Base64.encodeBase64((clientId + ":" +
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Use constants for all hard coded values

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 21]
```diff
@@ -0,0 +1,547 @@
...
+        return isOauthStateParamExists(request) && isOauth2CodeParamExists(request);
+    }
+
+    @Override
+    public String getContextIdentifier(HttpServletRequest request) {
+
+        String state;
+        try {
+            state = OAuthAuthzResponse.oauthCodeAuthzResponse(request).getState();
+            return state.split(",")[0];
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java:463]: Can this throw a arrayIndexOitofbound

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 22]
```diff
@@ -0,0 +1,547 @@
...
+    public String getName() {
+
+        return Oauth2GenericAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    protected void initTokenEndpoint(Map<String, String> authenticatorProperties) {
+
+        String tokenUrl = authenticatorProperties.get(Oauth2GenericAuthenticatorConstants.OAUTH_TOKEN_URL);
+        if (StringUtils.isEmpty(tokenUrl))
+            this.tokenEndpoint = getAuthenticatorConfig().getParameterMap()
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java:497]: Do not write inline if conditions. Not recommended as they are not readable

**Review by GDRDABARERA — CHANGES_REQUESTED** [review]

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java (hunk 21 above)
Comment (krishnilak) [reply components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java:463]: I added a seperate catch

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 23]
```diff
@@ -0,0 +1,553 @@
...
+
+            String state = stateToken + "," + Oauth2GenericAuthenticatorConstants.OAUTH2_LOGIN_TYPE;
+            OAuthClientRequest authzRequest = OAuthClientRequest.authorizationLocation(authorizationEP)
+                    .setClientId(clientId).setResponseType(Oauth2GenericAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE)
+                    .setRedirectURI(callbackUrl)
+                    .setState(state)
+                    .buildQueryMessage();
+
+            if (logger.isDebugEnabled()) {
+                logger.debug(String.join("Authorization Request",authzRequest.getLocationUri()));
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: logger.debug("Authorization Request: " + authzRequest.getLocationUri());

**Review by GDRDABARERA — APPROVED** [review]

**Pull request merged**